Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/DelegateActionListener.java

```diff
@@ -0,0 +1,176 @@
+import com.microsoft.azuretools.telemetrywrapper.TelemetryManager;
+import com.microsoft.tooling.msservices.components.DefaultLoader;
+import lombok.Getter;
+import lombok.Lombok;
+
+class DelegateActionListener extends NodeActionListener {
+
+    protected final NodeActionListener delegate;
+
+    private DelegateActionListener(NodeActionListener delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void actionPerformed(NodeActionEvent e) throws AzureCmdException {
+        delegate.actionPerformed(e);
+    }
+
+    @Override
+    public AzureIconSymbol getIconSymbol() {
+        return delegate.getIconSymbol();
+    }
+
+    @Override
+    public int getPriority() {
+        return delegate.getPriority();
+    }
+
+    @Override
+    public int getGroup() {
+        return delegate.getGroup();
+    }
+
+    static final class BackgroundActionListener extends DelegateActionListener {
+
+        private final String progressMessage;
+        private boolean cancellable;
+        private boolean conditional;
+
+        public BackgroundActionListener(NodeActionListener delegate, @NotNull String progressMessage, boolean cancellable, boolean conditional) {
+            super(delegate);
+            this.progressMessage = progressMessage;
+            this.cancellable = cancellable;
+            this.conditional = conditional;
+        }
+
+        @Override
+        public void actionPerformed(NodeActionEvent e) {
+            Runnable runnable = () -> {
+                try {
+                    delegate.actionPerformed(e);
+                } catch (AzureCmdException ex) {
+                    Lombok.sneakyThrow(ex);
+                }
+            };
+            AzureTask task = new AzureTask(e.getAction().getNode().getProject(), progressMessage, cancellable, runnable);
+            if (conditional) {
+                AzureTaskManager.getInstance().runInModal(task);
+            } else {
+                AzureTaskManager.getInstance().runInBackground(task);
+            }
+        }
+
+    }
+
+    static final class PromptActionListener extends DelegateActionListener {
+        private static final String PROMPT_TITLE = "Azure Explorer";
+        private static final String[] PROMPT_OPTIONS = new String[] {"Yes", "No"};
+
+        private String promptMessage;
+
+        public PromptActionListener(NodeActionListener delegate, @NotNull String promptMessage) {
+            super(delegate);
+            this.promptMessage = promptMessage;
+        }
+
+        @Override
+        public void actionPerformed(NodeActionEvent e) throws AzureCmdException {
+            boolean confirmed = DefaultLoader.getUIHelper().showConfirmation(promptMessage, PROMPT_TITLE, PROMPT_OPTIONS, null);
+            if (confirmed) {
+                delegate.actionPerformed(e);
+            }
+        }
+    }
+
+    static final class TelemetricActionListener extends DelegateActionListener {
```

Comment:
only telemetricActionListener will send telemetry when execute? What if I need to send telemetry in prompt actions?

Reply:
just need to implement both Temetricalble and Promptable interfaces.